recipe-ts-runtime/src/Recipe.ts

```diff
@@ -1,4 +1,139 @@
 import { Ingredient } from "./Ingredient";
+import { KeyedIngredient } from "./KeyedIngredient";
+
+export interface Segment {
+    domain: string | null;
+    recipe: Recipe;
+}
 
 export class Recipe extends Ingredient {
+    private readonly ingredients: Ingredient[];
+    private context: string | null = null;
+    private contextIngredient: KeyedIngredient | null = null;
+
+    public static prepare(...ingredients: Ingredient[]): Recipe {
+        return new Recipe(...ingredients);
+    }
+
+    public static context(context: string | KeyedIngredient, ...ingredients: Ingredient[]): Recipe {
+        const recipe = new Recipe(...ingredients);
+        recipe.context = (typeof context === "string") ? context : null;
+        recipe.contextIngredient = (context instanceof KeyedIngredient) ? context : null;
+        return recipe;
+    }
+
+    public getIngredients(): Ingredient[] {
+        return this.ingredients;
+    }
+
+    public getContext(): string | null {
+        return this.context;
+    }
+
+    public getContextIngredient(): KeyedIngredient | null {
+        return this.contextIngredient;
+    }
+
+    public segment(): Segment[] {
+        const segments: Segment[] = [];
+        const recipeStack: Recipe[] = [];
+        recipeStack.push(this);
+        this._segment(new Recipe(), recipeStack, null, segments);
+        return segments;
+    }
+
+    private _segment(currRecipe: Recipe | null, recipeStack: Recipe[], currDomain: string | null, segments: Segment[]) {
+        if (recipeStack[0].contextIngredient !== null) {
+            const contextIngredient = recipeStack[0].contextIngredient;
+            if (contextIngredient!.getDomain() !== currDomain) {
+                //copy recipe structure
+                let outerRecipe: Recipe | null = null;
+                let recipe: Recipe | null = null;
+                for (const r of recipeStack) {
+                    if (recipe === null) {
+                        outerRecipe = new Recipe();
+                        recipe = outerRecipe;
+                    }
+                    else {
+                        outerRecipe = Recipe.prepare(outerRecipe!);
+                    }
+                    outerRecipe.context = r.context;
+                    outerRecipe.contextIngredient = r.contextIngredient;
+                }
+
+                const segmented: Segment = {
+                    domain: contextIngredient!.getDomain(),
+                    recipe: outerRecipe!
+                };
+                segments.push(segmented);
+
+                currRecipe = recipe;
+                currDomain = contextIngredient!.getDomain();
+            }
+
+            currRecipe!.contextIngredient = contextIngredient;
+        }
+        for (const ingredient of recipeStack[0].ingredients) {
+            if (ingredient instanceof Recipe) {
+                let recipe = new Recipe();
+                recipe.contextIngredient = (<Recipe>ingredient).contextIngredient;
```

Comment:
This one's also weird... isn't typescript smart enough to realize `ingredient instanceof Recipe` because you're in an `if` that checked that?

Reply:
Actually, yes it is. This is just a remnant of me literally converting each line from Java to TypeScript.